File: src/org/heli/Camera.java
package org.heli;


import javax.media.opengl.GL;
import javax.media.opengl.GL2;
import javax.media.opengl.glu.GLU;

/** This class defines a camera in three space.
 * Future Optimizations: Animations, such as chase mode
 * Copyright 2015
 * @author Daniel LaFuze
 *
 */
public class Camera {
	/** This defines where the camera is located.
	 * 
	 */
	Point3D source;
	
	double orbitAltitude;
	
	double orbitRadius;
	
	private GLU glu;
	
	/** This defines where the camera is looking.  By default, it's facing the
	 * origin.
	 */
	Point3D target;
	
	/** This defines which direction is up.  It is intended to be a
	 * unit vector, though OpenGL isn't picky about the magnitude.
	 */
	Point3D upUnit;
	
	/** This specifies the field of view of the camera
	 * By default, it's 60 degrees.
	 */
	double fovDegrees;
	
	/** This specifies the near clipping distance, by default it's 1 unit.
	 * 
	 */
	double nearClip;
	
	int sceneWidth;
	
	int sceneHeight;
	
	/** This specifies the far clipping distance, by default it's 100 units.
	 * 
	 */
	double farClip;
	
	/** This is used for rotating (sines/cosines) -- in radians
	 * 
	 */
	double curAngle;

	public Camera(int trgX, int trgY, int trgZ, GLU theGLU) {
		// TODO: Compute source position such that default field of view
		// shows the whole thing... By default, camera source is up along
		// z axis
		orbitAltitude = 100.0;
		orbitRadius = trgX;
		source = new Point3D((double)(-trgX),(double)trgY,orbitAltitude);
		target = new Point3D((double)trgX, (double)trgY, (double)trgZ);
		upUnit = new Point3D(0.0,0.0,1.0);
		fovDegrees = 60.0;
		nearClip = 5.0;
		farClip = 1500.0;
		glu = theGLU;
		System.out.println("Camera at " + source.info() + " looking at " + target.info());
	}
	
	/** This constructor for a camera sets the defaults
	 * 
	 */
	public Camera(GLU theGLU) {
		source = new Point3D(0.0,0.0,5.0);
		target = new Point3D();
		upUnit = new Point3D(0.0,0.0,1.0);
		fovDegrees = 60.0;
		nearClip = 1.0;
		farClip = 100.0;
		glu = theGLU;
		sceneWidth = 100; // default, override with tellGL on resize
		sceneHeight = 100;
		curAngle = 0.0;
	}
	
	public void setTarget(Point3D inPoint)
	{
		target = inPoint;
	}
	
	/**
	 * This method allows the camera to move randomly
	 * @param radius
	 */
	public void wobble(double radius) {
		double deltaX = 2.0 * Math.random() - 1.0;
		double deltaY = 2.0 * Math.random() - 1.0;
		double deltaZ = 2.0 * Math.random() - 1.0;
		Point3D newPoint = new Point3D(source.x() + deltaX, source.y() + deltaY, source.z() + deltaZ);
		source = newPoint;
	}
	
	public void chase(Point3D newTarget, double minDistance)
	{
		if (minDistance < nearClip)
		{
			minDistance = nearClip;
		}
		target = newTarget;
		double actDistance = eyeDistance();
		if (actDistance > minDistance * 1.1)
		{
			approach(0.5 * (actDistance / minDistance));
		}
	}
	
	public void chase(Point3D newTarget)
	{
		target = newTarget;
		approach(1.0);
	}
	
	double eyeDistance()
	{
		double deltaZ = source.z() - target.z();
		double deltaY = source.y() - target.y();
		double deltaX = source.x() - target.x();
		double magnitude = Math.sqrt(deltaX * deltaX + deltaY * deltaY + deltaZ * deltaZ);
		return magnitude;
	}
	
	public void approach(double approachPercent) {
		double deltaZ = approachPercent / 100.0 * (source.z() - target.z());
		double deltaY = approachPercent / 100.0 * (source.y() - target.y());
		double deltaX = approachPercent / 100.0 * (source.x() - target.x());
		double magnitude = Math.sqrt(deltaX * deltaX + deltaY * deltaY + deltaZ * deltaZ);
		source.m_x = source.m_x - deltaX;
		source.m_y = source.m_y - deltaY;
		source.m_z = source.m_z - deltaZ;
	}
	
	public void orbit(double ticksPerRevolution) {
		if (ticksPerRevolution < 60.0)
		{
			ticksPerRevolution = 60.0;
		}
		curAngle += Math.PI /ticksPerRevolution;
		double deltaX = orbitRadius * Math.sin(curAngle);
		double deltaY = orbitRadius * Math.cos(curAngle);
		Point3D newSource = new Point3D(target.x() + deltaX, target.y() + deltaY, target.z() + orbitAltitude);
		source = newSource;
	}
	
	public void tellGL(GL2 gl, int w, int h)
	{
		sceneWidth = w;
		sceneHeight = h;
		tellGL(gl);
	}
	
	private void show()
	{
		System.out.println("Camera at: " + source.info());
		System.out.println("Camera Looking at: " + target.info());
	}
	
	public void tellGL(GL2 gl)
	{
		gl.glMatrixMode(gl.GL_PROJECTION);
		gl.glLoadIdentity();
		
		// Perspective
		if (sceneHeight == 0) // paranoid div/0 check
		{
			sceneHeight = 1;
		}
		double aspectRatio = (double) sceneWidth / (double) sceneHeight;
		glu.gluPerspective(fovDegrees, aspectRatio, nearClip,  farClip);
		glu.gluLookAt((float)source.x(), (float)source.y(), (float)source.z(),
				(float)(target.x()), (float)(target.y()), (float)(target.z()),
				(float)(upUnit.x()), (float)(upUnit.y()), (float)(upUnit.z()));
		
		gl.glMatrixMode(gl.GL_MODELVIEW);
		gl.glLoadIdentity();
	}
}

File: python/Camera.py
#!/usr/bin/python3

# (c) 2015-2024, A Beloussov, D. Lafuze

from panda3d.core import Vec3, Vec2
import random
import math

gMinDefault = 1.4

class HeliCamera:
    def __init__(self, trgX = 0, trgY = 0, trgZ = 0):
        self.orbitAltitude = 20.0
        self.orbitRadius = trgX
        self.source = Vec3(trgX,(-trgY/2),self.orbitAltitude)
        self.target = Vec3(trgX, trgY, trgZ)
        self.upUnit = Vec3(0.0,1.0,0.0)
        self.fovDegrees = 60.0
        self.nearClip = 5.0
        self.farClip = 1500.0
        self.curAngle = 0
        self.sceneWidth = 100
        self.sceneHeight = 100

    def setTarget(self, inPoint):
        self.target = inPoint

    def wobble(self):
        deltaX = 2 * random.randint(0,100) * 0.02 - 1
        deltaY = 2 * random.randint(0,100) * 0.02 - 1
        deltaZ = 2 * random.randint(0,100) * 0.02 - 1
        self.source = Vec3(self.source.getX() + deltaX, 
                           self.source.getY() + deltaY, 
                           self.source.getZ() + deltaZ)
        
    def approach(self,approachPercent):
        appr = (approachPercent / 100)
        deltaD = (self.source - self.target)
        deltaD = Vec3(deltaD.getX() * appr, deltaD.getY() * appr, deltaD.getZ() * appr)
        dist = deltaD.length()
        self.source = self.source - deltaD

    def chase(self, newTarget, minDistance = gMinDefault):
        if minDistance < self.nearClip:
            minDistance = self.nearClip

        oTrg = self.target
        oSrc = self.source
        self.target = newTarget
        above = 3

        if self.source.getZ() < (self.target.getZ() + above):
            self.source.setZ(self.target.getZ() + above)
        
        actDistance = (self.source - self.target).length()

        if actDistance > (minDistance * 1.04):
            self.approach(0.25 * (actDistance / minDistance))

    def orbit(self, ticksPerRev):
        if ticksPerRev < 60:
            ticksPerRev = 60
        self.curAngle = math.pi / ticksPerRev
        dX = self.orbitRadius * math.sin(self.curAngle)
        dY = self.orbitRadius * math.cos(self.curAngle)
        self.source = Vec3(self.target.getX() + dX,
                            self.target.getY() + dY,
                            self.target.getZ() + self.orbitAltitude)
        
    def getHpr(self):
        x2 = self.target.getX()
        y2 = self.target.getY()
        z2 = self.target.getZ()
        
        x1 = self.source.getX()
        y1 = self.source.getY()
        z1 = self.source.getZ()


        # Calculate yaw angle
        yaw = math.atan2(x2 - x1, y2 - y1)

        # Calculate pitch angle
        pitch = math.atan2(y2 - y1, x2 - x1)

        # Calculate roll angle
        roll = math.atan2(z2 - z1, math.hypot(x2 - x1, y2 - y1))

        print("yaw: ",yaw, ",pitch: ",pitch, ",roll: ",roll)

        res = Vec3(math.degrees(yaw),
                   math.degrees(pitch), #panda HeliCamera is looking at +Y
                   math.degrees(roll))

        return res


if __name__ == "__main__":
    camFlow = HeliCamera()
    camFlow.setTarget(Vec3(12,33,11))
    camFlow.approach(20)
    print(camFlow.source)
    camFlow.wobble()
    print(camFlow.source)
    camFlow.chase(Vec3(22,33,11))
    print(camFlow.source)
    camFlow.chase(Vec3(32,30, 10))
    print(camFlow.source)
    camFlow.orbit(30)
    print(camFlow.source)
    camFlow.source = Vec3(0,0,0)
    camFlow.target = Vec3(300,300,1)
    print(camFlow.getHpr())